File: Frontend/src/GameStore.Frontend/Services/BasketState.cs
using System.IdentityModel.Tokens.Jwt;
using GameStore.Frontend.Clients;
using GameStore.Frontend.Models;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Authorization;

namespace GameStore.Frontend.Services;

public class BasketState(
    IBasketClient client,
    AuthenticationStateProvider authenticationStateProvider)
{
    private EventCallback OnBasketUpdated { get; set; }

    private Task<CustomerBasket>? cachedBasket;

    public Task<CustomerBasket> GetBasketAsync()
    {
        return cachedBasket ??= FetchBasketAsync();
    }

    public async Task<CommandResult> AddItemAsync(BasketItem newItem)
    {
        var basket = await GetBasketAsync();

        basket.Items.Add(newItem);

        var result = await client.UpdateBasketAsync(basket);

        if (result.Succeeded)
        {
            // Invalidate the cache
            cachedBasket = null;

            // Notify subscriber
            await OnBasketUpdated.InvokeAsync();
        }

        return result;
    }

    public async Task<CommandResult> UpdateQuantityAsync(Guid id, int quantity)
    {
        var basket = await GetBasketAsync();

        var basketItem = basket.Items.Find(item => item.Id == id);

        if (basketItem is null)
        {
            // We will let the caller re-fetch the basket
            return new CommandResult(true);
        }

        basketItem.Quantity = quantity;

        var result = await client.UpdateBasketAsync(basket);

        if (result.Succeeded)
        {
            // Invalidate the cache
            cachedBasket = null;

            // Notify subscriber
            await OnBasketUpdated.InvokeAsync();
        }

        return result;
    }

    public async Task<CommandResult> RemoveItemAsync(Guid itemId)
    {
        var basket = await GetBasketAsync();

        basket.Items.RemoveAll(item => item.Id == itemId);

        var result = await client.UpdateBasketAsync(basket);

        if (result.Succeeded)
        {
            // Invalidate the cache
            cachedBasket = null;

            // Notify subscriber
            await OnBasketUpdated.InvokeAsync();
        }

        return result;
    }

    public void NotifyOnBasketUpdated(EventCallback callback)
    {
        OnBasketUpdated = callback;
    }

    private async Task<CustomerBasket> FetchBasketAsync()
    {
        var userId = await GetUserIdAsync();

        if (userId is null)
        {
            return new CustomerBasket() { CustomerId = Guid.Empty };
        }

        return await client.GetBasketAsync(userId);
    }

    private async Task<string?> GetUserIdAsync()
    {
        var authenticationState = await authenticationStateProvider.GetAuthenticationStateAsync();

        return authenticationState?.User?.FindFirst(JwtRegisteredClaimNames.Sub)?.Value;
    }
}

File: Frontend/src/services/BasketState.ts
import { CustomerBasket } from '../models/CustomerBasket';
import { CommandResult } from '../models/CommandResult';
import { BasketItem } from '../models/BasketItem';
import BasketClient from '../clients/BasketClient';

export class BasketState {
  private client: BasketClient;
  private userId: string | null;
  private cachedBasket: Promise<CustomerBasket> | null = null;
  private onBasketUpdated: (() => void) | null = null;

  constructor(client: BasketClient, userId: string | null) {
    this.client = client;
    this.userId = userId;
  }

  public async getBasketAsync(): Promise<CustomerBasket> {
    if (!this.cachedBasket) {
      this.cachedBasket = this.fetchBasketAsync();
    }
    return this.cachedBasket;
  }

  public async addItemAsync(newItem: BasketItem): Promise<CommandResult> {
    const basket = await this.getBasketAsync();
    basket.items.push(newItem);

    const result = await this.client.updateBasketAsync(basket);

    if (result.succeeded) {
      this.cachedBasket = null;
      this.notifyOnBasketUpdated();
    }

    return result;
  }

  public async updateQuantityAsync(id: string, quantity: number): Promise<CommandResult> {
    const basket = await this.getBasketAsync();
    const basketItem = basket.items.find(item => item.id === id);

    if (!basketItem) {
      return { succeeded: true, errors: [] };
    }

    basketItem.quantity = quantity;

    const result = await this.client.updateBasketAsync(basket);

    if (result.succeeded) {
      this.cachedBasket = null;
      this.notifyOnBasketUpdated();
    }

    return result;
  }

  public async removeItemAsync(itemId: string): Promise<CommandResult> {
    const basket = await this.getBasketAsync();
    basket.items = basket.items.filter(item => item.id !== itemId);

    const result = await this.client.updateBasketAsync(basket);

    if (result.succeeded) {
      this.cachedBasket = null;
      this.notifyOnBasketUpdated();
    }

    return result;
  }

  public setOnBasketUpdated(callback: () => void): void {
    this.onBasketUpdated = callback;
  }

  private async fetchBasketAsync(): Promise<CustomerBasket> {
    if (!this.userId) {
      return { customerId: '', items: [], totalAmount: 0 };
    }

    return await this.client.getBasketAsync(this.userId);
  }

  private notifyOnBasketUpdated(): void {
    if (this.onBasketUpdated) {
      this.onBasketUpdated();
    }
  }
}